### ISSUE TITLE
Admin panel: place beers on taplist

### ISSUE THREAD

**Issue by Skemp45** [RaspberryPints/RaspberryPints#40]
Admin panel: place beers on taplist

lets discuss this, I want to understand what you guys have in mind for the flow and functionality behind this. Once I have an understanding of how we want this to work I can start on it and once that's done all we need is the F.A.Q section then its just onto bug testing the admin panel for release.

Edited by Thadius 18 Jan 2014: Updating title and first line to jive with new issue/milestone tracking scheme.

**Comment by thadius856** [comment]
No problems. I know I outlined my vision before, but there were a ton of variables involved. We've actually already moved mostly in that direction.

Beers (recipes) and taps (faucets) should be segregated data. Beers/recipes on one table, what's active on tap on another table.

So how do we get there? Well, the Beer List already does the first half, essentially.

Rename `srm`, `og`, `fg`, `ibu` to `srmEst`, `ogEst`, `fgEst`, and `ibuEst`.
Remove `active` from the `beers` table. Add to the `tap` table instead.
Add `srmAct`, `ogAct`, `fgAct`, and `ibuAct` to the `tap` table.

The "est" values are what the recipe called for when we made the beer. But maybe we missed the gravity by a few points or messed up the hopping, so this batch is different. Nobody hits all 4 variables on even most batches.

On the taplist itself (index.php), it should use the Act variables if they're filled in. If they're null, it falls back to use Est.

Now for the Tap List table.

It should list however many taps we have installed. Each tap should have a dropdown list where we can pick from one of the beers on the ``beers` table. Imagine we're a year in and we have 50 recipes we've made. In the same row, boxes where we can fill in the Act values, if we desire.

If I'm unclear, email me a full size screenshot of the Beer List and Tap List and I'll draw it out in mspaint: thadius856@gmail.com

**Comment by JayUnt** [comment]
Does anyone remember the issue where we discussed the taps table?  I am going to create a script to make the table, but i wanted to make sure i got all of our changes.  Once you see the table, it might make more sense.

**Comment by JayUnt** [comment]
i guessi should have refreshed my screen before i committed that :)

**Comment by JayUnt** [comment]
Just a reminder, when a beer is added to the taps table, we have to make sure all of the other items in that table with that same tap number are set to inactive.

**Comment by thadius856** [comment]
> Just a reminder, when a beer is added to the taps table, we have to make sure all of the other items in that table with that same tap number are set to inactive.

Other than look odd having two identical tap numbers, does it break any functionality?

**Comment by JayUnt** [comment]
No it would work fine. We will be changing the beer query to really select
the active taps. So we would have a tap with two beers on it.

**Comment by thadius856** [comment]
Downloaded a copy of Balsamiq. Damn I love this program! Here's what I had envisioned, more or less.

![a](https://f.cloud.github.com/assets/6320180/1946201/02d3ffbe-7ff2-11e3-9211-7252c5285716.png)

Edit: Click to zoom.

**Comment by Skemp45** [comment]
That is not only beautiful but also very doable !!!!

**Comment by Skemp45** [comment]
Ok,

Thinking about it and let me know if this does not make sense because I may have had 1 too many union jacks.

When a user adds a beer on the beer list page should it auto set it to inactive and they will be activating it on the tap list page correct?

**Comment by thadius856** [comment]
Sorta correct. Active shouldn't even be a column on the `beers` table.

When a user adds a beer, it adds the beername to the drop down list on the Tap List page. Nothing more.

Edit: When the user performs initial setup, they specify their number of taps on the first line. It would auto-fill that many taps as ~Inactive~. When a user selects a beer for that tap, it saves beers(id) to tap(beer[i]) or somesuch.

**Comment by Skemp45** [comment]
So short of adding a beer what is the beer list page for? That's it?

**Comment by thadius856** [comment]
Yep. That's it. Rename it Add a Beer if you prefer. ;)

Edit: That's why I recommended Recipe List as a name for it. Leaves open the possibility of a Commercial Beers feature some day for commercial kegs or even bottles (RFID).

Leaves options for future. Think of the posibilities for optional columns.

Yeast?
Grain bill?
Hopping schedule?

Full recipe?

Import BeerSmith file?

**Comment by Boombrewer** [comment]
I like where this is going. Mainly because of its approach towards repeat brews. Most guys with multiple taps will really dig that aspect.

**Comment by JayUnt** [comment]
I also like the idea of having the beer's untappd id to show qr code to send the person right to the beer's page on untappd.  We should add that to the release features list. Not necessarily this release but for the future.

**Comment by thadius856** [comment]
Added to the list as "future".

**Comment by Skemp45** [comment]
ok guys, I added my changed to the admin panel personalize page. i added another config file called configp.php
in the /admin/includes. at the end I may try to reduce the # of configs to make it easy. but, you can now enable and disable columns and 4k monitor support all from one form.

**Comment by Skemp45** [comment]
Can someone also take the background.jpg and move a copy to admin/img and replace that background.jpg with our default one we want to use. this is where the restore default pulls from

**Comment by Skemp45** [comment]
What page will users be able to delete beers?

**Comment by Skemp45** [comment]
![taplist](https://f.cloud.github.com/assets/6318869/1949898/c893f768-8128-11e3-98e1-ba94b41b6d38.png)

**Comment by Skemp45** [comment]
what do you think so far?

**Comment by thadius856** [comment]
> Can someone also take the background.jpg and move a copy to admin/img and replace that background.jpg with our default one we want to use. this is where the restore default pulls from

Sure. One moment. See Commit 112 -> ed0f116f4b102ea7e7edfa21ace7303061d3582e

Not seeing many of your changes to the admin panel. Are you all sync'd up?

> what do you think so far?

Beautiful! Does it actually draw the number of taps you enter? :)

Beer Name might be a smidge on the narrow side, as far as columns go.

**Comment by thadius856** [comment]
> What page will users be able to delete beers?

The "Add a Beer" page in your screenshot seems most apt. Thought it might need a new name.

**Comment by Skemp45** [comment]
you dont see the changes?

**Comment by Skemp45** [comment]
should have been the added update_column.php file and a
update to personalize.php

**Comment by thadius856** [comment]
Yes, I see those, but not this update to the Tap List page.

**Comment by Skemp45** [comment]
ahh I am still working on that one as we speak

**Comment by thadius856** [comment]
Oh ok. Hehe :)

**Comment by Skemp45** [comment]
also, the beer name column sizes based off of what is in the drop down

**Comment by thadius856** [comment]
Cool!

**Comment by Skemp45** [comment]
we need to get the database structure setup and permanent so we dont have to keep going back and redoing. someone want to setup the db the way it needs to be?

**Comment by JayUnt** [comment]
I assumed you were doing it since you have them in the admin, but I can work on this too if you want.

**Comment by Skemp45** [comment]
I just dont want to set it up one way if you guys already had an idea on how you wanted it setup just go ahead with that and I will work around that

**Comment by JayUnt** [comment]
What is the keg type column supposed to show?  Im assuming we have preset values?  Will the user be able to add custom keg types?  if so we will probably want this to be another table, which is why I'm asking.  Even if the custom keg types are going to be another release.

**Comment by JayUnt** [comment]
Also this profile table looks to be redundant.  We should really put this stuff in the config table, otherwise it will get confusing.  Ill go ahead and do it, but i just wanted to give you the heads up.

**Comment by Skemp45** [comment]
ok

**Comment by Skemp45** [comment]
once this is done I will go in and update the areas that need updated based off of moving from profile to config and then commit those changes.

**Comment by Skemp45** [comment]
Jay,

any suggestions on how I should logically approach this? Trying to figure out how to take what the user inputs here and making it limit the amount of tap spots there are. I was thinking maybe having a value called taplimit in the config table and have the number they put in there update the value of that option in the config table. Then using the variable $taplimit to set the number. IDK kind of stuck on this one. 
![update](https://f.cloud.github.com/assets/6318869/1955161/621f1f82-81e4-11e3-876c-a567de8fbf17.png)

**Comment by JayUnt** [comment]
Yeah, putting a maxTaps in the config table is a good idea.  When this number changes, we should update the taps table to disable all taps that are a greater number.  and if it is increased, we can either update the latest row for that tap number or just make them insert a new row at that time.  I'm thinking a new one might be easier/less confusing for the user.

**Comment by JayUnt** [comment]
I've committed my changes to the schema.  Take a look and let me know if you have any changes/questions.

I have not updated the front end to use the updated fields, but i will do that later.  So if you do update your database, the front end will break.  We will have to make the changes anyway to pull from the taps table, so this will be a good time to do it.

A couple of more things I've noticed that I changed, but want you guys to see it just so we can prevent it in the future: 
- The users table was set to use ENGINE=MyISAM, when the rest of the tables are using ENGINE=InnoDB.
- The tables that are using AUTO_INCREMENT had incremental values set, some were jumping by 4 one was by 15.  There is no need for that because we just want them all to increment by 1.
- Config was using _ instead of camel case for the column names (this was my doing, whoops!)
- I added a createdDate and a modifiedDate column to each of the tables.  We need to make sure to insert the current timestamp in to the created date and modified date on creation of the row, and insert update the modified date when we update the row.

**Comment by Skemp45** [comment]
OK, so i dont understand why we got rid of profile as thats what the admin panel used for the user photo url and for the header text. 

also note Users was changed to users

**Comment by JayUnt** [comment]
ah yes, i should have mentioned the Users > users. sorry.  I wanted to keep it consistent with the other table name formatting.

To me, the columns in the profile table were configurations.  when i first saw there was a profile table, i wasn't sure what it was a profile for.  I thought it was a user's profile, but then thought why wouldn't those appear in the users table?  After looking at the rows, i realized it just different config values for the system.  Not everything in the config table has to be able to be updated by the user. Its really system config values, where we allow the users to update the ones we allow.  

Also if we ever add a new configuration, its a lot cleaner to just add a new row into the table then it is to alter a table to add this column.

Unless you think we are going to give the user the ability to select one of our preset "skins", i would say we not user the profile table.  And even if we do allow preset "skins", i would suggest we have a skins table and then the config table have a value that points to that skin's id.

let me know if that clarified it for you.

**Comment by Skemp45** [comment]
ahh I see what your saying, just throws off alot of the script i had :( going to have to figure out how to rewrite it

**Comment by Skemp45** [comment]
I had 
<?php
            $sql="SELECT \* FROM profile";
$result=mysql_query($sql);

$rows=mysql_fetch_array($result);?>

<h2>Change Header Text</h2>

```
        <form method="post" action="update_header_text.php"><br />
         <b>Header text</b><br />
        <input type="text" value="<? echo $rows['header_text']; ?>" name="header_text"> &nbsp <br />
        <input type="submit" class="btn" name="Submit" value="Submit">
        </form>
```

but thats not going to work as its now a config_Name and now the actual text is config_Value

**Comment by thadius856** [comment]
I'm a bit confused here. Profile table? I don't see that word in the schema at all.

If you want to pull the options for the config page straight from the config table, just add another column to it to signify if you want the value user editable. That is, add...

```
  `userEditable` tinyint NOT NULL,
```

...and add 0/1 to each column in the data dump below.

On the config page, just add a WHERE to select only the 1's.

---

Unless I've missed the mark here. I may be completely misunderstanding the situation.

**Comment by thadius856** [comment]
> What is the keg type column supposed to show? Im assuming we have preset values? Will the user be able to add custom keg types? if so we will probably want this to be another table, which is why I'm asking. Even if the custom keg types are going to be another release.

Different keg types. This is mainly so should I ever get bored, I can make stroke/mask images for each keg type and have the taplist display the appropriate one.

Also, I may use the keg type later for calibrating a "full keg". That is, a user weighs their full keg before adding to the keezer and types the weight into the Taplist admin page. If I had the weight of an empty keg of each type, I could easily estimate how much beer was in their keg. And that could set their kegStart column or whatever we have it set to now.

I'll be committing the sample keg sizes shortly.

Edit: So no, the user won't be able to add custom keg sizes. But I will be including enough kegTypes out-of-the-box that they shouldn't want for anything.

**Comment by Skemp45** [comment]
also note 

config was changed from 
CREATE TABLE `config` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `config_name` varchar(50) NOT NULL,
  `config_value` longtext NOT NULL,
  `display_name` varchar(65) NOT NULL 

to 

CREATE TABLE `config` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `configName` varchar(50) NOT NULL,
  `configValue` longtext NOT NULL,
  `displayName` varchar(65) NOT NULL,

**Comment by JayUnt** [comment]
Yeah sorry. I didnt keep track of every change, but i probably should have.

The userEditable in the config makes sense.  That way we don't have to hard code the editable options in the admin, unless we want to hardcode it so so we can set the order of the inputs on the screen.

**Comment by Skemp45** [comment]
no worried I was able to adjsut script to pull header text and admin logo  url without issue

**Comment by Skemp45** [comment]
I had to do this  
$sql="SELECT configValue FROM config WHERE id = 8";
$result=mysql_query($sql);

$rows=mysql_fetch_array($result);

**Comment by Skemp45** [comment]
<?php
            $sql="SELECT configValue FROM config WHERE configName ='headerText'";
$result=mysql_query($sql);

$rows=mysql_fetch_array($result);?>

**Comment by Skemp45** [comment]
changed it to this that way id is not given

**Comment by JayUnt** [comment]
I would suggest you use the ConfigNames when making the query. That will
keep everything using that and not hardcoding the string in different
spots.  I've already added the value to it, so all you have to do is use
it.  And if you aren't using it in the other admin sections, we should
change them over.  Makes everything cleaner and if we change a configName,
its easier to find throughout the site.

On Mon, Jan 20, 2014 at 1:11 PM, Skemp45 notifications@github.com wrote:

> changed it to this that way id is not given
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/RaspberryPints/RaspberryPints/issues/40#issuecomment-32783138
> .

## 

Jason Unterman

**Comment by Skemp45** [comment]
not sure what you mean

**Comment by Skemp45** [comment]
```
    <?php
            $sql="SELECT configValue FROM config WHERE configName ='headerText'";
$result=mysql_query($sql);

$rows=mysql_fetch_array($result);?>
<h2>Change Header Text</h2>
            <form method="post" action="update_header_text.php"><br />
             <b>Header text</b><br />
            <input type="text" value="<? echo $rows['configValue']; ?>" name="header_text"> &nbsp <br />
            <input type="submit" class="btn" name="Submit" value="Submit">
            </form>
```

i need to spit out the config value as the way it is setup in the db the configValue is what text displays for "beers on tap"

**Comment by JayUnt** [comment]
If you look at index.php and look at the <thead> area, you will see that im using ConfigNames::ShowTapNumCol inside $config.  ConfigNames is a class that was created to store all of the configNames from the table.  This way im only hardcoding the column names in one spot and using that throughout the site.

**Comment by thadius856** [comment]
@Skemp45 

Looks like all your code is getting processed by GitHub Flavored Markdown.

You can wrap code with three left quotes, similar to [code][/code] on bbCode.

I'll update the post above*\* to illustrate it for you.

**Comment by Skemp45** [comment]
Beautiful

**Comment by Skemp45** [comment]
```
<?php
abstract class ConfigNames
{
    const ShowTapNumCol = 'showTapNumCol';
    const ShowSrmCol = 'showSrmCol';
    const ShowIbuCol = 'showIbuCol';
    const ShowCalCol = 'showCalCol';
    const ShowAbvCol = 'showAbvCol';
    const ShowKegCol = 'showKegCol';    
    const UseHighResolution = 'useHighResolution';
    const LogoUrl = 'logoUrl';  
    const HeaderText = 'headerText';    
}
```

I see what you mean I added these here. 
how would I go about echoing those into an input then?

**Comment by thadius856** [comment]
@Skemp45 For multi-line code, the open and close should be on a newline, or you'll get weird rendering.

Example:

```
'''
<?php
abstract class ConfigNames
{
    snip
}
'''
```

Not:

```
'''<?php
abstract class ConfigNames
{
    snip
}'''
```

I had to use single quotes above in the example so it would render. But you should be using left quotes (`). :)

**Comment by JayUnt** [comment]
You need to make sure that the includes/config_names.php is included in the page.  then you can just use

```
ConfigNames::ShowTapNumCol
```

to call the value.  You can see it being used in index.php.

**Comment by JayUnt** [comment]
```
$sql="SELECT configValue FROM config WHERE configName ='".ConfigNames::HeaderText."'";
$result=mysql_query($sql);
$headerText=mysql_fetch_array($result);
```

Then you can use this to output it:

```
$headerText['configValue']
```

I wonder if we should turn that 'configValue' into that same class or another class, so we dont have to hardcode that either :/

**Comment by Skemp45** [comment]
nice I got that to work :D ty

**Comment by Skemp45** [comment]
I cleaned up the admin panel repo. deleted the files not needed and limited it down to 2 config files. So I would suggest re downloading the admin!

**Comment by Skemp45** [comment]
conn.php and configp.php

**Comment by JayUnt** [comment]
Good to hear it.  Its tough with this small of a project to really enforce the rules that should be enforced.  Im used to using Visual Studio where if you tried to use ConfigNames::ASDASDASD, it wouldn't compile and you fix it before you commit, but with notepad++ and php, I don't have that luxury.

And with the smaller projects you think, "its one spot, ill just hard code the string", but once it grows it you can lose control fast.  Theys why I'm trying to keep things consistent from the beginning.

**Comment by thadius856** [comment]
Agree on the nature of hardcoding. Even with small projects, I try to reference includes and make formatting as extensible as possible.

This translates well into other office activities: spreadsheets, presentations, folder structures, etc.

**Comment by Skemp45** [comment]
Thanks for showing me that, that does make it much easier and cleaner

**Comment by JayUnt** [comment]
I updated the front end to now pull from the taps table.  So index should now work for everyone again.

I also added a script to the test_data.sql that populate the taps table.  it takes all of the beers from the beers table and creates a tap using its data. 

Im not a huge fan of how the query is in the top of index.php.  it really should be either in its own file, or better yet we can even make a db view for it.  I had to write it in mysql and then turn it into a string.  that doesnt allow easy debugging.  Ill take a look more at that tonight.

**Comment by thadius856** [comment]
Hrm. Dropped my db, re-imported schema successfully. test_data, not so successfully.

```
Error
SQL query: 

-- --------------------------------------------------------
--
-- Put all beers into the `taps` table
--
INSERT INTO taps(  `beerId` ,  `kegTypeId` ,  `tapNumber` ,  `active` ,  `ogAct` ,  `fgAct` ,  `srmAct` ,  `ibuAct` ,  `startAmount` ,  `createdDate` , `modifiedDate` ) 
SELECT b.id, kegType.id AS kegTypeId,  '1' AS tapNumber, 1 AS active, b.ogEst AS ogAct, b.fgEst AS fgAct, b.srmEst AS srmAct, b.ibuEst AS ibuAct, kegType.maxAmount AS startAmount, NOW( ) AS createdDate, NOW( ) AS modifiedDate
FROM  `beers` AS b, (


SELECT * 
FROM kegtypes
ORDER BY Id
LIMIT 1
) AS kegType/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;

MySQL said: 

#1064 - You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */' at line 16 
```

**Comment by JayUnt** [comment]
that is fixed.  forgot a ; with the taps query.

**Comment by JayUnt** [comment]
hopefully you didnt do an another update yet because i just committed another change.  instead of having the select taps query in the index.php i made a database view for it.  That makes it easier to dev and debug.  So you will need to rerun everything again :)

**Comment by thadius856** [comment]
Just did. But it's like a 3 second process with TortoiseGit anyway.

**Comment by thadius856** [comment]
You've got some typos still in test_data. Pushing a commit to fix them now, but there's still more errors to iron out.

**Comment by JayUnt** [comment]
That's weird, they both ran fine for me right before it committed. If you
want me to look into them, can you post the errors?

**Comment by thadius856** [comment]
Sure. Using the test_data.sql you committed:

```
Error
SQL query: 

-- --------------------------------------------------------
--
-- Put all beers into the `taps` table
--
INSERT INTO taps(  `beerId` ,  `kegTypeId` ,  `tapNumber` ,  `active` ,  `ogAct` ,  `fgAct` ,  `srmAct` ,  `ibuAct` ,  `startAmount` ,  `createdDate` , `modifiedDate` ) 
SELECT b.id, kegType.id AS kegTypeId,  '1' AS tapNumber, 1 AS active, b.ogEst AS ogAct, b.fgEst AS fgAct, b.srmEst AS srmAct, b.ibuEst AS ibuAct, kegType.maxAmount AS startAmount, NOW( ) AS createdDate, NOW( ) AS modifiedDate
FROM  `beers` AS b, (


SELECT * 
FROM kegtypes
ORDER BY Id
LIMIT 1
) AS kegType;

MySQL said: 

#1146 - Table 'raspberrypints.kegtypes' doesn't exist 
```

This one is pretty clear. SELECT \* FROM kegtypes. Correct table is kegTypes. Fixed, then re-ran. It went in fine.

Committing now.

**Comment by thadius856** [comment]
Odd side-effect. Tapnumber circles lost their valign: middle and all show "1" now.

![untitled](https://f.cloud.github.com/assets/6320180/1959203/5c7f86cc-8228-11e3-9b08-0d4f0d14eb18.png)

**Comment by thadius856** [comment]
Oddly, beer_list.php now displays completely blank. tap_list.php does not show any beers, even though I have 10 beers loaded.

**Comment by thadius856** [comment]
Odd, I just uploaded 10 beers in test_data, dropped my db and reloaded.

Now they display on the frontend in an order different than the appear in test_data.

![untitled](https://f.cloud.github.com/assets/6320180/1959372/59f08cc2-822c-11e3-8c57-a43f393158e8.png)

**Comment by JayUnt** [comment]
This has been fixed. you just need to clear the taps table and then run that query again. It was a bug in the query.  Its not exactly how i want it, but i cant think of a better way to set the tapNumber.  Bu you should be good to go.

And the order issue was wacky because the tap numbers were all set to 1.

**Comment by JayUnt** [comment]
Skemp, 

Are you writing all of these sql queries that are in the admin section? Or do you have some other way that you are coding, that these queries are being built for you automatically?

The reason i ask is because i'm curious to your thoughts on switching over to use an ORM instead of just putting sql queries throughout the pages?  Ive used Doctrine in the past, which has worked but is a bit large, especially for this application which is going to go on a pi.  But I'm looking at other ones that are smaller and will make the CRUD a lot simpler.  What do you think?

I also saw that the header and left menu bar code is on every single page, instead of just being in includes.  If it is automatically generated, what are you using?  If not, we should really put them in includes, so the code is in only one spot.

**Comment by Skemp45** [comment]
We are working now on cleaning up and simplifying the admin. that includes making changes on one page to reflect all pages. also using assets, and functions. Trying to follow best practices and making this as clean as possible.

**Comment by Skemp45** [comment]
I have one of the Admins from my work assisting me with following best practices and cleaning up the admin panel, His name is Ethan and you can see his repo here, https://github.com/twansolo/RaspberryPints.

**Comment by Skemp45** [comment]
and to answer your question JayUnt I have been writing them all.... UGH

**Comment by thadius856** [comment]
Ugh-ly! 

_See what I see there? *nudge nudge_*

**Comment by Skemp45** [comment]
hahaha Yeah Agreed, its about to get beautiful up in here though.

**Comment by JayUnt** [comment]
sounds good.  if you need help, let me know.  im staying away from the admin, to not step on your toes.

**Comment by thadius856** [comment]
I'll dip a toe in admin for proofreading later, but I too am trying to stay away.

Right now completing the instructions and FAQ are my priorities. And huge time sinks both.

**Comment by Skemp45** [comment]
Jay, 

Feel free to work on the tap list page. Also, I am sorry that I wasn't really able to work on this today. I was really busy at work and then had my monthly AZ homebrewer meeting. I should be able to put more time in tomorrow. In the mean time, if you need anything let me know. 

Shawn

PS. Here is a picture of the turnout at the local AZ homebrewer meeting (ASH).. pretty cool! 
![homebrewermeeting](https://f.cloud.github.com/assets/6318869/1971059/71676644-831e-11e3-8af8-064d1409af1a.jpg)

**Comment by Skemp45** [comment]
500 members strong!

**Comment by Skemp45** [comment]
Set all navagation and footers using php includes so they can be updated on one file and reflect across all pages

**Comment by thadius856** [comment]
If opening or or closing issues with a milestone assigned, please update
the milestone description (to-do list) as well.

**Comment by Skemp45** [comment]
Also, I ordered my raspberry pi this morning and check this out
![yess](https://f.cloud.github.com/assets/6318869/1976396/dddea98e-8389-11e3-9f11-39cf57707410.png)
!!!!!! It will be here before I get home!!!!

**Comment by JayUnt** [comment]
Sounds good. ill work on the taps page.

**Comment by JayUnt** [comment]
Skemp,

I started working on the taps list and detoured a bit from what you were doing. I just wanted to run what ive done by you before i continue on my path.  So far i have the update number of taps working, and it shows the taps with the selected beer.

As an overall of what i did.  I made a couple of function php files in admin/includes.  One generic function, one for beer related functions, one for tap related functions and one for kegType related functions.  tap_list.php includes whatever php files it will need.  these functions are really database managent function, like updating the number of taps, getting a list of beers, etc.

Another thing i did different, was instead of the tap_list.php submiting to a different page that does the sql and redirecting back, i just submit to the same page and then call one of the functions, currently it only calls the update number of taps, but it will eventually call a function that updates the taps.

One thing I didnt see and was wondering about is, are you doing any validation on the beers import?  Like a beer's og is a number and not letters?

Let me know of what you think about how i was going about this.

Thanks

**Comment by Skemp45** [comment]
I think...... what you are doing is awesome and the direction i wanted to take it so PLEASE keep on the path you are down. I like the idea of submitting from one page. Keep up the amazing work jay. and during testing I was not checking for numbers etc. but planed to set that on production

**Comment by JayUnt** [comment]
Awesome! :)

I do front end development for a living and we use MVC as our application structure.  I'm just trying to mimic that sort of structure, but without the object, since i have never used objects like this in php. But keeping the database functions in their own files, if we ever wanted to change from mysql to another database, it would be a matter of including a different file with the same functions.

The next thing i really would like to look into is having actual models for all beer, taps, etc.  Ill probably look into that when i add the tap inputs to the form.

**Comment by JayUnt** [comment]
i made a few more changes to what I had.  I turned the different object function files into classes and moved them into a managers folder.  I did this so is more of a scope and will prevent conflicts between the different files having the same function name.  

So we could have a $beerManager->Save($beer) and a $tapManager->Save($tap).  They both have a Save() function, but there will be no conflicts about it.

Also renames the functions.php to html_helper.php and made it a class too.  Since it was the only function in that file and its more specific now.  We can add a functions.php back later if we want.

**Comment by thadius856** [comment]
The beer_main.php page now loads for me. However, it shows no beers. I have 10 `beers` table. If I try to add one, it just refreshes the form (or fails to submit the data).

Same deal with tap_list.php. Zero taps showing. If I try to update to 10, it just refreshes.

Reeks of permissions issue. What am I doing wrong / not have set up?

**Comment by Skemp45** [comment]
I must say the tap_list page is looking great. Nice work Jay

**Comment by JayUnt** [comment]
Thanks Skemp.

I have a usability question for you guys.  So we have estimated stats (eg, fg, etc.) for each beer,  When the user changes a tap from one beer to another, should i just automatically overwrite what is in the tap's stats textboxes?  Do you think that could get annoying to the user, to lose what they put in, if they picked the wrong beer at first, updated the stats and then had to change the beer again?  

Should we do a confirmation message, asking if they want the stats overwritten automatically? Or should we just do it automatically?

P.S. for some reason the kegType table, the name was an int, so i changed it to text.  

I also created an object for taps, with properties and whatnot.  I will do the same for kegType and we should also do it for beers.

**Comment by JayUnt** [comment]
And i have another question.  

So lets say i setup a tap of Darth Vader.  A month goes bye and i use it up.  I then want to put another keg of Darth Vader on that tap to replace it.  How does the system know its a new keg?  Especially if the keg was from the same batch, so it has the same og, fg, etc.

Another problem, is what happens if i add a tap to a keg and then realize i messed up the ibu.  How do we know this is just an update to a tap and not a new tap?

One solution is to have a submit with each tap. When they click that it will update the existing tap.  There will also be a checkbox labeled "New Keg",  if that is checked, we will make it a new row in the db, and disable the previous one.  

A problem with this is that if they change beers and dont check the "new keg" checkbox, it will just overwrite the existing tap.  We could popup a confirm when they change their beer, asking if its a new different.  butt that would interfere with my last question about automatically populating the data.

Idea?

**Comment by thadius856** [comment]
> > I have a usability question for you guys. So we have estimated stats
> > (eg, fg, etc.) for each beer, When the user changes a tap from one beer to
> > another, should i just automatically overwrite what is in the tap's stats
> > textboxes? Do you think that could get annoying to the user, to lose what
> > they put in, if they picked the wrong beer at first, updated the stats and
> > then had to change the beer again?
> 
> Should probably be whenever you pick a beer from the list, it writes the
> estOg/Fg/Ibu/etc into the actOg/Fg/Ibu/etc boxes.

Not at home so I can't see it, but... we don't have an estAbv/actAbv, do
we? ABV should be calculated from OG - FG.

> Should we do a confirmation message, asking if they want the stats
> overwritten automatically? Or should we just do it automatically?

I wouldn't bother. I don't think it would be annoying.

> P.S. for some reason the kegType table, the name was an int, so i changed
> it to text.

Cool, thanks.

> I also created an object for taps, with properties and whatnot. I will do
> the same for kegType and we should also do it for beers.

Greek to me. One more time, in English?

**Comment by JayUnt** [comment]
So when the beer type is changed,  i will just replace the stats with the beer's stats.

**Comment by thadius856** [comment]
> And i have another question. So lets say i setup a tap of Darth Vader. A
> month goes bye and i use it up. I then want to put another keg of Darth
> Vader on that tap to replace it. How does the system know its a new keg?
> Especially if the keg was from the same batch, so it has the same og, fg,
> etc.

My vision is that when you load a keg, if you want to utilize flow meters,
you enter the starting amount in the keg. After each pour, we decrement.
This amount should be the right-most column on the tap_list page, IIRC. Not
at home, but pretty sure I put a starting amount column on there in the
mockup.

> Another problem, is what happens if i add a tap to a keg and then realize
> i messed up the ibu. How do we know this is just an update to a tap and not
> a new tap?

Not sure what this isn't a problem. Wouldn't the field just update
actIbu for that tap? Maybe I'm misunderstanding.

> One solution is to have a submit with each tap. When they click that it
> will update the existing tap. There will also be a checkbox labeled "New
> Keg", if that is checked, we will make it a new row in the db, and disable
> the previous one.

Why not just re-write the old tap? You want to save a history or something?

> A problem with this is that if they change beers and dont check the "new
> keg" checkbox, it will just overwrite the existing tap. We could popup a
> confirm when they change their beer, asking if its a new different. butt
> that would interfere with my last question about automatically populating
> the data.
> Uhhhh... what?

**Comment by JayUnt** [comment]
The ABV is not being saved in the db, it will be calculated.

The objects is not really going to be noticeable by the user, but will make coding easier.  I more said it for Skemp to see what I was doing.

But for a quick explanation, instead of storing the tap's values in array, in which its easy to assign ibu to ibuEst by mistake, i made a Tap object.  So if you try to assign it to ibuEst, it will break when you get to the page, so we can catch it easier.  Otherwise it will just insert null into the database, and we might not realize.

We dont want to overwrite the tap with the new beer, because then we would lose all of our tap history.  The idea is that if its a new beer, you update the tap in the admin, and we would set the previous one to inactive and then create a new tap row.  This way we can track how many times a beer was on tap, what the pours were for that tap, etc.  if we just update row with the new beer's id, then all the pours from tap 1 will merge, even if they were different beers.

My idea is to have another column in the row, that is for the user to tell us that its a new beer, so we can go ahead and create a new tap row.

**Comment by JayUnt** [comment]
And as for the flow meters, that will just update the pours table. We have to keep the fill level alone so we know how much of the keg was used for the graph.

**Comment by JayUnt** [comment]
Another solution will be to have a button on each row labeled "Close Out Tap",  if they click it, we clear out the form row for them, and we know that we can start a new tap row in the db.

If they don't click it though, they can get poor analytics because all beers on that tap will be merged.

Ill work on a mockup like you have above with these changes.  Maybe that will make it more clear.

**Comment by thadius856** [comment]
>  The idea is that if its a new beer, you update the tap in the admin,
> and we would set the previous one to inactive and then create a new tap
> row. This way we can track how many times a beer was on tap, what the pours
> were for that tap, etc. if we just update row with the new beer's id, then
> all the pours from tap 1 will merge, even if they were different beers.

If you didn't want to know how many times each beer was on tap, this
wouldn't be so difficult. Because the pours table contains almost all of
the other analytics.

How about this.

User selects a beer recipe from the dropdown. They fill in the vitals (ibu,
srm, etc). Then at the far right row, we do a check. (bear with my
shorthand)

if (kegLvL)
     display kegLvl in a box (non-editable)
else
     display an empty editable box and a Save button. Save this value
to BOTH kegStart and kegLvl.

Only problem is then you couldn't manually edit the beer level without
digging into myPhpAdmin.

> My idea is to have another column in the row, that is for the user to
> tell us that its a new beer, so we can go ahead and create a new tap row.

With what I described above, you could have a button "Swap Keg" or "Replace
Keg" or "New Keg" or somesuch.

Or even a "Keg Kicked!" button that sets kegLvl to 0, then deducts the
difference from kegStart (because clearly you had more or less beer than
you originally thought, or your flow meters aren't calibrated).

**Comment by Skemp45** [comment]
Ok, so I think when putting a new beer on tap it should over write the old numbers. 

Each beer should have its own submit button I feel that way users don't try to do a bunch at once and mess up.

If there is a darth vader on tap but empty and they want to just out another keg on we should have a renew button.

As for pop up are your sure etc. I love these, working in the tech industry I realized one thing, PEOPLE ARE CLICK HAPPY. Let's take precautions and use them to make people read!!!

**Comment by thadius856** [comment]
> And as for the flow meters, that will just update the pours table. We
> have to keep the fill level alone so we know how much of the keg was used
> for the graph.

The solution above fixes this problem.

My only worry with not decrementing some value after each pour is the lag
caused by loading the last x number of transactions to check the level.

Say I have 10 beers on tap (I do). And each 5 gal keg takes 53 glasses to
empty (they do). That means I need to look at the last 530 pours (on
average) to see all the transcations for that keg. Or maybe more, if it's a
lesser-used keg.

**Comment by thadius856** [comment]
> Each beer should have its own submit button I feel that way users don't
> try to do a bunch at once and mess up.

I could go either way on this one.

> If there is a darth vader on tap but empty and they want to just out
> another keg on we should have a renew button.

"Renew". I like this.

> As for pop up are your sure etc. I love these, working in the tech
> industry I realized one thing, PEOPLE ARE CLICK HAPPY. Let's take
> precautions and use them to make people read!!!

Yeah, I see it all day. People don't read error messages either. They get
to the word "error" and their eyes glaze over.

**Comment by JayUnt** [comment]
So we can have two amount on the tap.  the fill amount, which is actually called startAmount in the db already. And then we have a current amount, which we set to the start amount on tap creation, but then subtract from each time a pour is made.  Then we wouldn't need to calculate it each time they load the page.  But this doesn't really solve our "when do we consider it a new keg" question. 

Having a "Keg Kicked" button would do that.  It would tell us to mark the current tap row in the db inactive, so we know to start a new tap row when they set a new keg, even if it the same beer recipe.  So if they want to change kegs, they would have to click this button and then we would clear it out.  This would also be used for people who aren't going to use flow meters.  It will still allow us to know that they are putting on a different keg, as opposed to just fixing a beer recipe mistake.  We would be able to tell how many times they tapped this beer before, which i think is good for the analytics.  We would just have to tell them to use it, so we can track this info.  We could use the confirm modal if they change the beer name, to see if they want to kick it, or if they are fixing a mistake.

With the ppl with the flow meters, we could disable the beer dropdown for a tap once a pour is created.  This would force them to kick the keg if they want to change the keg.

**Comment by thadius856** [comment]
I like where you're going with this.

There's really three reasons they'd swap a tap:
1. The keg kicked. They're putting in something different.
2. The keg kicked. They're putting in another keg of the same thing.
3. The keg didn't kick, but they want to swap for whatever reason.

For #1, the Kicked! button works.
For #2, we could go with a Kicked! button and then have them re-submit the same recipe. Or, we could give them a Renew! button.
For #3, we could allow them to pause a keg. Or not, I don't really care. :)

**Comment by JayUnt** [comment]
Im good wtih #1

i think for #2, we just have them kick it, it gets cleared, and they re-select the beer again.  No need to have another 
button.

And #3 they just edit it and its edited, however they want.

Do we want to lock the selected beer recipe if there are any pours for that beer?  Or even let them change the beer type too?  My reasoning is that if there is a pour, odd are the beer is right and that there shouldn't be any changed.  I'm trying to prevent them from setting up a beer, and then when its done, they just change the beer type from the list.  I'm not sure how we do that with non-flow sensor people, maybe it available to edit for a certain amount of time after adding. Then when that time i up, if they want to change the beer, they have to kick it and then create a new one.

**Comment by thadius856** [comment]
Hrm. What did you plan to deliver analytics for people not using flowmeters?

Temperature vs time, maybe?

If they don't want flow meters, I don't see why we should even bother trying to track volume for them.

**Comment by JayUnt** [comment]
Another thing we can do is have a read only version of the tap list.  Then in each row, there is a "tap a keg" link if there is nothing tapped. This would give them a single tap form.  

If there is something tapped, we can have a "fix mistake" link which will allow them to edit it, "keg kicked" link that will make the current level 0 and remove keg from tap, and even a "change keg" link which lets them not set the current level to 0, but still inactivate that tap.  This would force them to explicitly click on the action they want to do.  As opposed to allowing them to edit it and we are hoping they tell us what they are doing.

**Comment by JayUnt** [comment]
yeah we dont need to track volume for non-flow meters, but it would be nice to see how much they started with.  We can also track how many times a beer was tapepd.

The analytics will be nice for the more automated activitied, like flow metered pours, and who poured it (eventually :) )  The more analytic we can think of, the more fun that section will be, even if its not that important of data.

**Comment by thadius856** [comment]
> The beer_main.php page now loads for me. However, it shows no beers. I have 10 beers table. If I try to add one, it just refreshes the form (or fails to submit the data).
> 
> Same deal with tap_list.php. Zero taps showing. If I try to update to 10, it just refreshes.
> 
> Reeks of permissions issue. What am I doing wrong / not have set up?

Any word on this? I'm still hard broke.

**Comment by JayUnt** [comment]
We could also tell them what beers they tend to tap the different seasons. Could help the keep track of when they should brew their summer beer, because it goes on and gets kicked during the summer fast.

**Comment by thadius856** [comment]
Ooooh, that's a good idear.

Ties in well with another feature I have planned for wayyyyyyy down the road. :dancer: 

But it's a secret. Shhhh. ;)

**Comment by JayUnt** [comment]
oooh i like secret features.  Im trying to keep as much historical data as we can, so we can have back data for fun charts and whatnot.  

I haven't touched the beers page. i've been focusing on the taps page.  What were your thoughts on my readonly taps list, and forcing the user to select, "Edit tap", "kick tap", etc? as opposed to having all of the edits on one page?

on a side note.  Where do you get all of these cool icons? i only see the ones that start with ":".

**Comment by Skemp45** [comment]
I love analytics

**Comment by thadius856** [comment]
https://github.com/blog/1289-emoji-autocomplete

**Comment by thadius856** [comment]
:dancer: was supposed to be a :D tho

**Comment by JayUnt** [comment]
Haha

**Comment by thadius856** [comment]
So... is there something wrong with my SQL user accounts? Users table?

**Comment by JayUnt** [comment]
There are no errors avoiding up? Just a refresh? 

I knew you guys were seeing up different users for the front end and the admin, but I just an using beer/beer for both. And he has full rights.

**Comment by thadius856** [comment]
Hrm. Lemme try that. beer//beer is only SELECT for me.

**Comment by thadius856** [comment]
```
GRANT SELECT , 
INSERT ,

UPDATE ,
DELETE
```

No change.

**Comment by thadius856** [comment]
Also added beer//beer to table users. No change.

It's odd to me that I have 10 beers in beers, yet zero show on either page. And I can't add any taps.

**Comment by JayUnt** [comment]
Just to clarify, the talks page isn't done. The only pay off the form that works is the update radio number. The taps will list, but the submit to edit the taps doesn't work yet. I was holding off on finishing it until after we talked about the changes.

**Comment by JayUnt** [comment]
As for the beer page, ill check to see if it with four me in a bit, but I haven't gone there in a while.

**Comment by JayUnt** [comment]
Here is an example to what i was talking about with a read only view of the tap list.  the style isnt full there but you can get the idea.  They user will have to explicitly tell us that they are changing the info vs adding a new keg to that tap.  We could get rid of the replace keg, and force them to "kick keg" and then "tap a beer".  replace keg is just a shortcut to kick it and start a new one.  kick keg will just remove it and make that row have "tap a keg" link like the last row.

If they click "tap a beer", "update tap info" or "replace keg", they will be sent to a single tap form, like the beer one.

I think its a lot cleaner and easier for the user, so I'm going to continue on this path, unless anyone has issues with it.

![2014-01-24 14_59_21-raspberrypints](https://f.cloud.github.com/assets/35524/1998352/5f726d2e-8532-11e3-82d6-e2993f099588.png)

**Comment by Skemp45** [comment]
Ah I like that a lot, it would help reduce user error and a lot of confusion

**Comment by thadius856** [comment]
I like that too.

Once you get it working,  let me know. I'd like to take a stab at
appearance and readability.

**Comment by thadius856** [comment]
Can we make it into a table with SRM, IBU, OG, FG etc etc sub-headers as only showing once at the top?

**Comment by JayUnt** [comment]
we sure can.

@Skemp45 ,  Do you have any issues with me updating jquery?  im not sure where else its used, so i dont know what else the update could break.

**Comment by Skemp45** [comment]
That's grand, go ahead with the jquery  update!!!!!!!

**Comment by JayUnt** [comment]
Alright guys,  big changes to the tap stuff.  Its not 100% tested, there is no validation yet and the its a very basic style for now (including not changing the headers that you requested yet), but i wanted to commit to so you guys can see it.

I would have tested more last night when i was developing it, but my laptop battery dies, so i took that as a sign to go to sleep.  Ill test/add validation most likely tonight, but i might be able to get some in during the day too.

**Comment by thadius856** [comment]
Still hard broke. At least the page loads now. :\

![untitled](https://f.cloud.github.com/assets/6320180/2002353/d2875aba-85ee-11e3-9bd3-4fa27e54b41c.png)

![untitled](https://f.cloud.github.com/assets/6320180/2002354/dbb6eba0-85ee-11e3-8327-94156ab09a4d.png)

**Comment by JayUnt** [comment]
I made some changes to the taps page that might have fixed parts. I don't
have the php shortcode enabled so I can't use just <? But have to use <?php
to start the php code. It's possible you gave the same php config, so my
changes fixed done if them.  Do the tap pages work for you?

**Comment by thadius856** [comment]
> It's possible you gave the same php config, so my
> changes fixed done if them.

Not sure what you're asking.

`short_open_tag` is set to `On` in /etc/php5/apache2/PHP.ini

> Do the tap pages work for you?

The one in the screenshot above? No.

Does the front end `index.php` work? Yes.

**Comment by JayUnt** [comment]
Yeah that was what I was asking about.  The short_open_tag.

Can you view the source of the admin taps list, and see if there looks too
be any error messages in the html? You can copy it to here if you want me
to look through it.

**Comment by thadius856** [comment]
Sure.

```

<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<title>RaspberryPints-Login</title>
<link href="styles/layout.css" rel="stylesheet" type="text/css" />
<link href="styles/login.css" rel="stylesheet" type="text/css" />
<!-- Theme Start -->
<link href="styles.css" rel="stylesheet" type="text/css" />
<!-- Theme End -->

</head>
<body>
    <div id="logincontainer">
        <div id="loginbox">
            <div id="loginheader">
                <a href="../" style="text-decoration:none;"><h1><font color="#2DABD5">RaspberryPints Login</h1></font></a>
            </div>
            <div id="innerlogin">
                <form name="login" action="includes/checklogin.php" method="POST">
                    <p>Enter your username:</p>
                    <input type="text" class="logininput" name="myusername" placeholder="Login Name" />
                    <p>Enter your password:</p>
                    <input type="password" class="logininput"  name="mypassword" placeholder="Password"/>

                    <input type="submit" class="loginbtn" value="Log In" /><br />
<img src="img/lock.png" height="50" width="50">
                    <p><a href="reset_account.php" title="Forgoteen Password?">Forgotten Password?</a></p>
                </form>
            </div>
        </div>
    </div>
</body>
</html>
```

**Comment by JayUnt** [comment]
What happens when you run this query on the database?

`SELECT * FROM `config` WHERE configName = 'numberOfTaps';`

It doesn't look like you have that config row.  if you dont, run this query and then refresh the page

`INSERT INTO `config`(configName, configValue, displayName, createdDate, modifiedDate) VALUES('numberOfTaps', '0', 'Number of Taps', NOW(), NOW());`

Then you should see a 0 in the Number Of Taps textbox.  You can then change it to whatever number you want and hit submit.  THHHHEEEEENNNNN you should see that number of tap rows, possibly all needing to have a beer attached to them.

**Comment by thadius856** [comment]
> `SELECT * FROM `config` WHERE configName = 'numberOfTaps';`

MySQL returned an empty result set (i.e. zero rows). ( Query took 0.0058 sec )

> `INSERT INTO `config`(configName, configValue, displayName, createdDate, modifiedDate) VALUES('numberOfTaps', '0', 'Number of Taps', NOW(), NOW());`

1 row inserted. 
Inserted row id: 9 ( Query took 0.0337 sec )

**Comment by thadius856** [comment]
> Then you should see a 0 in the Number Of Taps textbox. You can then change it to whatever number you want and hit submit. THHHHEEEEENNNNN you should see that number of tap rows, possibly all needing to have a beer attached to them.

Yup, worked. Now I see 10 on the tap list. Beer list is still empty.

**Comment by JayUnt** [comment]
excellent.  I haven't touched the beer list, but i can do the same thing to it as im doing with the taps, when I'm done.  I know Skemp was working on that section, but it would be nice to upgrade the code to match the taps one.  Ive just been working off of the test data beers.

**Comment by thadius856** [comment]
Added an update statement to the test data.

Really odd I didn't have the numberOfTaps row at all. I nuked everything and re-imported several times. Confirmed that schema.sql is building it out correctly. _shrug_

Committed -> 8954ca76094266f034fd0550c9148c5adb3199b1

**Comment by JayUnt** [comment]
there should be an insert in the schema file.  Though i might have added it recently :)

**Comment by Skemp45** [comment]
the beer list page does not list all of the beers only has the add a beer form, I switched this when we were making all the changes. I have not been able to do anything as of late as Ive been sick. Cons of working in a big office building, 1 person gets it EVERYONE gets it!

**Comment by thadius856** [comment]
Just tried out TortoiseGit's version of `git blame`.

Shows that line was added in 52b03256afbd9f25b9bab17fe7682dafc1b5fc95. 22 Jan 14.

So, yeah, recent. I haden't nuked raspberrypints in the few days because of work.

**Comment by thadius856** [comment]
> the beer list page does not list all of the beers only has the add a beer form, I switched this when we were making all the changes.

Oh ok. We'll need a list/delete functionality at the minimum for launch, preferably list/edit/delete.

With the delete button, I'd like to not actually delete the row from `beers`, but set a show/view value to 0. If we were to let users delete recipes, it may wreck havoc with stats down the road. For example:

I brew Beer A and put it on tap twice. Then I decide I no longer like Beer A, so I delete it. Now the stats page shows I tapped two kegs of "Null Value". Mmmm, delicious.

A few days later, I add the recipe for Beer C. It re-uses Beer A's old primary key. Now, I've tapped "Beer C" twice according to the stats page, when it hasn't even finished fermenting yet.

**Comment by thadius856** [comment]
>  I have not been able to do anything as of late as Ive been sick. 

No worries. Rest up.

> Cons of working in a big office building, 1 person gets it EVERYONE gets it!

Tell me about it. I work on a several thousand person campus and have people come in for customer service. Workstations are hot-desked for 24/7/365 operations. Sickness spreads like crazy here.

Luckily, I found that once I stopped biting my nails, started washing my hands more often during the course of the day, and started running 2-3 times per week, I don't get sick. Haven't been sick in about 2 years now, even when my whole office was sick a few weeks ago. _knock on wood_

Then again, three of us had to run a 9-man shop while they were out. So we didn't exactly skate away from it.

**Comment by Skemp45** [comment]
A: AHHh good point!!!

B: I am totally going to brew a beer called "Not Null"

C: I think we should rename the New Keg button to Replenish Keg

**Comment by thadius856** [comment]
I like your previous choice of "Renew".

**Comment by Skemp45** [comment]
Yeah either one, "New" just seems like new as in different beer.

**Comment by JayUnt** [comment]
Agreed with no delete.  Just call it inactivate and set active to false.

**Comment by JayUnt** [comment]
The reason i chose to use New Keg, is because Replenish Keg means we are making an assumption that you are going to update the tap with the same style of beer.  New keg, just means its a different keg, with the same style or not.  Its easy enough to say New Keg and then just select the same beer from the dropdown.

If we want to assume that they are going to replace a tap with a new keg of the same beer each time, we can, but i don't think that is really needed.

But if both of you want to use  Renew Keg to Replenish Keg, I can make that change and have the pervious beer's beer type pre-selected with its default stats.

**Comment by thadius856** [comment]
Ah, that makes sense. But only if when you hit "New Keg" it changes to the "Inactive" entry automatically.

**Comment by JayUnt** [comment]
Exactly, New Keg really is a Kick Keg automatically followed by a Tap A Keg.

**Comment by thadius856** [comment]
The more I think about the kegs on the taplist, the more I think it needs to be split out. Hear me out.

If we had a table `keg`, you could enter stats about your kegs. Perhaps:

Keg #1 is a 5 gallon corny keg that weighs 5.264 lbs when empty.
Keg #2 is a 5 gallon pin lock keg that weighs 4.987 lbs when empty.
Keg #3 is a 15.5 gallon Sanke keg that weighs (whatever a sanke keg weighs!) when empty.

If when putting a beer on tap, we had the user select a keg from table `keg` (via dropdown, just like they select a beer), it would make flow meter calibration (down the road) a whole lot easier.

Why? We already know what the keg weighs when empty. And the user just told us the keg weights 42.64 lbs, so we must have 4.321 gallons of beer in it. Easy!

Input?

**Comment by JayUnt** [comment]
Do different pin lock kegs weight different amounts, or are they really the same? If they are the same, we could just add weight to the kegType take. 

The only problem I have with a kegs table is having to keep track of which keg is which. If I have 10 pin locks knowing which is which would be annoying. 

What if we added empty weight to kegTypes? Then we give the user an admin screen to add/edit/delete keg types. Then if they want to add 10 pin locks with specific weights, they can.

By if you think it will make it easier for flow calibration or it will give them a better way to fill a start amount, by putting a full weight and we calculate the amount, let's do it. Get people used to the more setup needed now, instead of adding it later, is a good way to go.

**Comment by thadius856** [comment]
I've weighed several of my kegs, and they do weigh slightly different.

How do you track keg #s? Well, I'll be applying vinyl cutouts to mine, but I bet even a Sharpie would work.

I like where's you're going with the weight entry, though. It's like our estOg vs actOg. If we give users the opportunity to weigh each keg, even if they decide not to weigh them individually, they could just add 8 kegs with the default weight. That means I gotta add that column to the kegTypes table. _off to do that_

Since we already have a `kegTypes` table, perhaps the name `kegInv` or `kegInventory` would be more apt.

**Comment by JayUnt** [comment]
Yeah I guess just putting a sticker/vinal/sharpie on the keg is not a big
deal.

I actually prefer the kegs table name and keep the kegType table name as
is. You own a bunch of kegs and they each have a keg type.

You know what? It could be a good analytic too. Being able to see how often
certain kegs are used vs other ones. So you can maybe switch up which ones
you use, to spread it the wear and tear.  I always seem to go back to this
analytics stuff I guess :-)

I'm sold!

**Comment by thadius856** [comment]
Once a nerd, always a nerd. :P

**Comment by JayUnt** [comment]
:wink: :beers:

**Comment by thadius856** [comment]
Hey, there's a beers! Never noticed.

Writing the table now. Having trouble thinking of what else we'd want in it other than keg type and number/name.

Do you envision storing the keg usage stuff in `kegs` or `pours` or somewhere else for statistics?

**Comment by Boombrewer** [comment]
Yeah, but nerds make much cooler toys and gadgets.  We're all nerds with great hobbies!

**Comment by JayUnt** [comment]
We will want the empty weight and maybe max volume.

And we will want to change the kegTypeId in taps to be kegId and kegs will have a kegTypeId.  We will be able to track all we need with those associations.

**Comment by JayUnt** [comment]
And if course an is that's auto incremented.

**Comment by JayUnt** [comment]
*id

**Comment by thadius856** [comment]
Of course.

**Comment by Skemp45** [comment]
Sorry again I wasn't able to help much this weekend, I feel bad. Starting to get over this cold, drank a 75 minute tonight to help ;)!! Followed by 2 emergen c's lol

**Comment by thadius856** [comment]
Went and inventoried all of my kegs that not currently in the keezer. Found a couple other fields to add for them. Many are purely informational and "for fun", but still no hurt having them.

Added the `kegs` table. Also created a `kegStatuses` table to describe the different states a keg can be in (e.g., serving, primary, secondary, bulk aging, etc.).

Unfortunately I introduced a syntax error that prevents successful import of schema. Can't seem to locate it, but it seems to be near line 166 from what I got from the interpreter. Can you guys take a look to see if you can find anything obvious? Probably a comma or parenthesis or something.

Committed, so anybody downloading tonight is going to get a broken version. 36d6e40cfe87847a65dfdeb527b5889364b10c68

**Comment by thadius856** [comment]
> Sorry again I wasn't able to help much this weekend, I feel bad. Starting to get over this cold, drank a 75 minute tonight to help ;)!! Followed by 2 emergen c's lol

No worries, man. It's only Saturday night, so maybe you'll feel better tomorrow. ;)

Though probably not. Keep a glass of water nearby the nightstand. Your throat is gonna be killing you in the morning.

**Comment by Boombrewer** [comment]
Skemp,

Can you give me view-only access to the site so I can evaluate the how-to process?  If I can figure it out on my Mac, any idiot with a keyboard should be able to do it.

My username is boombrewer.

Thanks,
Casey

**Comment by Skemp45** [comment]
What's your email Casey

**Comment by Boombrewer** [comment]
nomadic.jester@gmail.com

I just signed up so I'm not sure if the account is actually up and running yet.  Haven't gotten the email yet.... didn't expect you to be so 'johnny on the spot'

hopefully it works.

**Comment by thadius856** [comment]
Then you must not have seen the code so far if you didn't expect that.

IT pros tend to be pretty quick on the email trigger. Our lives depend on
it. :)

**Comment by Boombrewer** [comment]
I guess I was just hoping he'd had a few beers since it was saturday night and wasn't watching email so closely.

ok... so the glass of stout out of the new keg might have been affecting my thinking... don't judge me.  I just installed my nitrogen system tonight and got to pour my first creamy stout.  I'm still in a state of delirium and bliss.

**Comment by thadius856** [comment]
I know the feeling. I just finished force carbing a keg of IPA (my fav
style), specifically Two Hearted, featuring 100% of my favorite hop
(Centennial).

Third glass now. Mmm, can't get enough.

**Comment by Boombrewer** [comment]
I'm not a huge IPA fan (mainly due to the overwhelming hop flavor) but they're growing on me.  I'm kicking around brewing my buddy's Red Simcoe recipe.  I just can't decide whether or not to pull the trigger.  Tomorrow I keg my first lager so I might need something quick to replace a fail.

**Comment by thadius856** [comment]
I'm not super big on IPAs either, because they wreck palletes.

This one is nicely balanced at about 52.6 IBU to 1.055 OG (0.95 BU:GU, so it's ever so slightly malt-forward). Most of the flavor comes from Centennial in dry hop, which gives a nice floral aroma and taste. Very little pucker to this one. If there was a lady's version of an IPA, it'd be either this or all-Citra.

**Comment by Boombrewer** [comment]
Hmm... now I'm intrigued.  Would you mind sharing the recipe?  If you don't want to send it to all, shoot it direct to nomadic.jester@gmail.com

**Comment by thadius856** [comment]
Here ya go man:

http://www.homebrewtalk.com/f69/bells-two-hearted-ale-clone-close-they-come-91488/

My second fav is an all-citra by a local brewery Lincoln Brewing Company. All citra, IIPA. About $8/btl tho.

**Comment by Boombrewer** [comment]
Thanks.  I like citra quite a bit as well.  You haven't tried cloning it?  at $8/btl, I'd be interrogating a brewmaster or seriously trying to figure out how to replicate it.

My buddy really likes citra hops.  I think he has a nice IPA that uses all citra.  Want me to track it down for ya?

**Comment by thadius856** [comment]
Nah, I only go to their tap room occasionally. Usually getting a DD is the problem. Wife isn't super receptive to the idea of her not drinking.

Here we go. 95 on BeerAdvoce. World-class.

http://beeradvocate.com/beer/profile/23200/78377/?ba=xmorgan

**Comment by Boombrewer** [comment]
That is a delicious looking brew.

**Comment by thadius856** [comment]
Whoops. Forgot to add estEmptyWeight to kegs table. Will do so in the
morning.

**Comment by Skemp45** [comment]
ok, got a chance to look at the updated to the admin. are we looking for a different page for each of these. 

Show/hide columns
headers
logo
background.
Because having 4 links to the same page seems redundant and makes it so A LOT is going on on that left panel.

**Comment by Skemp45** [comment]
Casey,

I created a user with the user boombrewer and your password was sent to your email. The login url is
http://raspberrypints.com/site/wp-admin

**Comment by Skemp45** [comment]
Closing thread starting new, 200 comments is actually causing load time lol.
